Added a security module

Fat-Frumos opened the pull request:
- Implemented an authentication service with controller levels
  - jwt token provider created
  - Jwt authorization filter added
  - security configuration added

Spring JPA data added
  - created a token repository

Updated controller integration tests
  - authentication controller tests added
  - modify integration tests with jwt token

Fat-Frumos committed:
Added a security module
  - Implemented an authentication service with controller levels
  - jwt token provider created
  - Jwt authorization filter added
  - security configuration added

Spring JPA data added
  - created a token repository

Updated controller integration tests
  - authentication controller tests added
  - modify integration tests with jwt token

Fat-Frumos committed:
Updated JWTProvider
 - refresh token added
Created permission tests for user, admin, guest roles
Added role entity
Postman collection was updated

Fat-Frumos committed:
Added a refresh token
Created a custom logout
Removed CookiesUtils
Updated access rights verification tests:
 - added CRUD-check for the administrator role,
 -  added CUD-check for user authentication

sslichnyi commented on Module#04/model/src/main/java/com/epam/esm/entity/Role.java:
```diff
@@ -0,0 +1,59 @@
+@AllArgsConstructor
+@Table(name = "roles")
+public class Role implements Serializable {
+    private static final long serialVersionUID = 3218292427916480827L;
```
serialVersionUID is redundant here because serialization is not used

Fat-Frumos replied:
> serialVersionUID is redundant here because serialization is not used

> serialVersionUID has been removed

sslichnyi commented on Module#04/model/src/main/java/com/epam/esm/entity/User.java:
```diff
@@ -77,13 +90,14 @@ public class User implements Serializable {
-            + "(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9]"
-            + "(?:[a-z0-9-]*[a-z0-9])?",
-            message = "{invalid.email}")
     private String email;
```
@Email validation annotatoin can be used from javax validation library

Fat-Frumos replied:
> @Email validation annotatoin can be used from javax validation library

> Added @Email annotation to the email field to ensure that the format of the email address is checked.

sslichnyi commented on Module#04/repository-spring-jpa/src/main/java/com/epam/esm/repository/impl/UserRepoImpl.java:
```diff
@@ -0,0 +1,77 @@
+     */
+    @Override
+    public List<User> getAllBy(final Pageable pageable) {
+        return userRepository.findAll(pageable)
```
if I am not wrong findAll() method returns ArrayList by default

Fat-Frumos replied:
> if I am not wrong findAll() method returns ArrayList by default

> PagingAndSortingRepository returns a Page of entities meeting the paging restriction provided in the Pageable object.

sslichnyi commented on Module#04/security-service/src/main/java/com/epam/esm/security/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,101 @@
+                                    response.sendRedirect("/api/token/login");
+                                })
+                                .and()
+                                .formLogin(formLogin -> formLogin.defaultSuccessUrl("/certificates"));
```
If I am not wrong formLogin() method is used in MVC with jsp pages or thymleaf, it will redirect to HTML page in stead to redirect to endpoint. In REST there is no  such thing as redirect, this is why HATEOAS is used

Fat-Frumos replied:
> If I am not wrong formLogin() method is used in MVC with jsp pages or thymleaf, it will redirect to HTML page in stead to redirect to endpoint. In REST there is no such thing as redirect, this is why HATEOAS is used

> Yes, you are right, this is used to redirect to the login page that the framework generates. This has been removed

sslichnyi commented on Module#04/security-service/src/main/java/com/epam/esm/security/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,83 @@
+                        .map(token -> !token.isExpired()
+                                && !token.isRevoked())
+                        .orElse(false);
+                if (provider.isTokenValid(jwt, userDetails) && isValid) {
```
it is better to swap  provider.isTokenValid(jwt, userDetails) and isValid.
it will validate isValid variable and if it is false then  provider.isTokenValid(jwt, userDetails) will not be triggered

Fat-Frumos replied:
> it is better to swap provider.isTokenValid(jwt, userDetails) and isValid. it will validate isValid variable and if it is false then provider.isTokenValid(jwt, userDetails) will not be triggered

> Swapped the order of validation, in the case, if isValid is false, the isTokenValid method will not be triggered, improving efficiency.

sslichnyi commented on Module#04/security-service/src/main/java/com/epam/esm/security/service/AuthenticationService.java:
```diff
@@ -0,0 +1,235 @@
+     *
+     * @return The default Role object.
+     */
+    private static Role getRole() {
```
it must be renamed to getUserRole or any other, it is not clear which role is User, Admin...

Fat-Frumos replied:
> it must be renamed to getUserRole or any other, it is not clear which role is User, Admin...

> Renamed the method to getDefaultUserRole(), making it clearer that it returns the default role for the user

sslichnyi commented on Module#04/security-service/src/main/java/com/epam/esm/security/service/AuthenticationService.java:
```diff
@@ -0,0 +1,235 @@
+        return AuthenticationResponse.builder()
+                .username(user.getUsername())
+                .expiresAt(Timestamp.from(now()
+                        .plusMillis(accessToken)))
```
accessToken is not understoodable name, Looks like it is object but it milis, please rename

Fat-Frumos replied:
> accessToken is not understoodable name, Looks like it is object but it milis, please rename

> accessToken has been renamed accessTokenExpirationMillis

sslichnyi commented on Module#04/security-service/src/main/java/com/epam/esm/security/service/OAuth2Service.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+
+public class OAuth2Service extends OidcUserService {
```
what is purpose of this service?

Fat-Frumos replied:
> what is purpose of this service?

> The purpose of the OAuth2Service class, is to provide functionality related to OAuth 2.0 authentication and user management. (Optional task: Use Oauth2 as an authorization protocol)

sslichnyi commented on Module#04/web-app/src/main/java/com/epam/esm/config/WebMvcConfig.java:
@EnableWebMvc @Configuration class is used two times, why?

sslichnyi commented on Module#04/web-app/src/main/java/com/epam/esm/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    @PostMapping("/refresh")
+    public ResponseEntity<AuthenticationResponse> refreshTokens(
+            final HttpServletRequest request,
```
HttpServletRequest is overhead to be used. it is better to use @Header("Authorization") annotation

Fat-Frumos replied:
> HttpServletRequest is overhead to be used. it is better to use @Header("Authorization") annotation

> modified signature refreshTokens to @RequestHeader("Authorization") String authorizationHeader

sslichnyi commented on Module#04/web-app/src/main/java/com/epam/esm/controller/UserController.java:
```diff
@@ -26,6 +26,7 @@
  */
 @RestController
 @RequiredArgsConstructor
+//@PreAuthorize("hasRole('ADMIN')")
```
uncommet it or remove please

Fat-Frumos replied:
> uncommet it or remove please

> added @EnableMethodSecurity to enable @PreAuthorize("isAuthenticated() && hasRole('ADMIN')")

Fat-Frumos committed:
Split the logic in the login service into signup and login
Created getAuthenticationResponse for the generic AuthenticationService response
Changed permissions for user and guest

Fat-Frumos committed:
Added updating, deleting services for entity user and order for admin permission
Updated tests for service, DAO layers

Fat-Frumos committed:
Modify permission for user
Added tests integration
Separeted User and Security UserDetails
Updated messages from server
